sycl/include/CL/sycl/ONEAPI/group_algorithm.hpp

```diff
@@ -190,6 +190,23 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
         !cl::sycl::detail::is_native_op<T, BinaryOperation>::value,
     T>;
 
+template <typename T> struct is_native_function_object : std::false_type {};
+
+template <template <typename> class BinaryOperation, typename T>
+struct is_native_function_object<BinaryOperation<T>> {
```

Comment:
Based on the spec I would have expected
```suggestion
template <template <typename> class T> struct is_native_function_object : std::false_type {};

template <template <typename> class BinaryOperation>
struct is_native_function_object<BinaryOperation> {
```
And the usage to be `is_native_function_object<plus>` not `is_native_function_object<plus<>>`. Why do we want the later?

And even if we take a template instantiation rather than the template as argument, why should we check the argument of the template? As you wrote, because of `void` the test isn't sufficient anyhow. Therefore why check at all? And not remove lines 198-200 and only check `is_native_op`?

Reply:
Currently spec (released 2020 Provisional) says nothing about this trait except "It should be". I believe that we still need `T` type to make sure we know the identity for that. Imagine that you have `std::plus` and `int`. Obviously the identity is 0.  But if you have ``std::plus`` and `class A` you don't know the identity for that. That's why I think having just non instantiated template is not sufficient.

Reply:
I don't think that works, at least not as you've written it.  The template arguments are the same there, so the second case isn't a template specialization.  I also encountered compiler errors trying to forward a template template parameter like that -- I couldn't find a way to write things that way without the compiler complaining that `BinaryOperation` needed a template argument.

I agree that if we can figure out how to check that the functor is one of `plus`, `multiplies` etc without looking at the type of the template argument it's a nicer solution.

Reply:
What works is:
```c++
template<template<class>class Op>
struct is_native_function_object : detail::is_native_op<void, Op<void>> {};
```
Is there a reason to check for anything other than `void`?

Reply:
> Is there a reason to check for anything other than void?

One potential use-case for this trait is going to be inside of things like oneDPL, as in the simplified pseudo-code below:
```c++
template <typename InputIt, typename BinaryOperation>
std::iterator_traits<InputIt>::value reduce(InputIt first, InputIt last, BinaryOperation binary_op) {
  if constexpr (is_native_function_object_v<BinaryOperation>) {
    /* implementation can leverage SYCL work-group and sub-group reductions */
  } else {
    /* implementation must provide its own work-group and sub-group reductions */
  }
}
```

If `is_native_function_object` only checks `void`, would you expect this to fail?

Thinking about it, this usage might be a good demonstration of why the usage should be `is_native_function_object<plus<>>` and not `is_native_function_object<plus>`.

Reply:
> If is_native_function_object only checks void, would you expect this to fail if the caller of reduce provided std::plus<float>()?

This has nothing to do with "checks `void`". This would fail for any `BinaryOperation`. Because independent of `float`/`void` it would always be a type (/template instantiation) and not a template.

> Thinking about it, this usage might be a good demonstration of why the usage should be is_native_function_object<plus<>> and not is_native_function_object<plus>.

I agree. This is an important usage. And given that the API which takes a type can be used in both cases easily, but the other only in the one case easily, we should use the `is_native_function_object<plus<>>` syntax.